server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java

```diff
@@ -140,8 +131,20 @@ public SegmentReplicationTargetService(
      */
     @Override
     public void beforeIndexShardClosed(ShardId shardId, @Nullable IndexShard indexShard, Settings indexSettings) {
-        if (indexShard != null) {
+        if (indexShard != null && indexShard.indexSettings().isSegRepEnabled()) {
             onGoingReplications.cancelForShard(shardId, "shard closed");
+            latestReceivedCheckpoint.remove(shardId);
+        }
+    }
+
+    /**
+     * Replay any received checkpoint while replica was recovering.  This does not need to happen
+     * for primary relocations because they recover from translog.
+     */
+    @Override
+    public void afterIndexShardStarted(IndexShard indexShard) {
+        if (indexShard.indexSettings().isSegRepEnabled() && indexShard.routingEntry().primary() == false) {
```

Comment:
nit: Do we need `indexShard != null` check here ?

Reply:
This is invoked through a composite listener [here](https://github.com/opensearch-project/OpenSearch/blob/9bf99b4a63de0e14811c0aeab0e9f854591f72ed/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L733) from within an initialized IndexShard's updateShardState, where the shard can't be null.